### PR TITLE
add DxfSpline -> Svg conversion

### LINKED ISSUE #5: DxfSpline to SVG conversion is not supported

### PR DESCRIPTION
**Pull request by brettfo**
@sampletext32 Can you see if this does what you expect?  Right now it only supports splines of degree 3 because those easily map to cubic Bezier curves; lowering the degree of more complex curves is currently beyond my mathematical ability.

Fixes #5.

### CODE REVIEW

**Review by sampletext32 — APPROVED** [review]
Seems good. Will run some tests and give you a feedback.

File: src/IxMilia.Converters/Spline2.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                // get next set of values
+                var values = result.KnotValues.Skip(offset * expectedIdenticalKnots).Take(expectedIdenticalKnots).ToList();
+
+                if (values.Count == 0 && result.KnotValues.Count() % expectedIdenticalKnots == 0)
+                {
+                    // done
+                    break;
+                }
+
+                var exepctedValue = values[0];
```
Comment (sampletext32) [inline src/IxMilia.Converters/Spline2.cs]:
```suggestion
                var expectedValue = values[0];
```

File: src/IxMilia.Converters/Spline2.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                throw new InvalidOperationException("Invalid combination of knot value count, control point count, and degree.");
+            }
+
+            // knot values must be non-descending and [0.0, 1.0]
+            var lastKnotValue = 0.0;
+            foreach (var kv in _knotValues)
+            {
+                if (kv < lastKnotValue)
+                {
+                    throw new InvalidOperationException("Knot values must be non-descending.");
```
Comment (sampletext32) [inline src/IxMilia.Converters/Spline2.cs]:
```suggestion
                    throw new InvalidOperationException("Knot values must be ascending.");
```

File: src/IxMilia.Converters/Spline2.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        private void ValidateValues()
+        {
+            if (_controlPoints.Count < Degree + 1)
+            {
+                throw new InvalidOperationException("There must be at least one more control point than the degree of the curve.");
+            }
+
+            if (_knotValues.Count != _controlPoints.Count + Degree + 1)
+            {
+                throw new InvalidOperationException("Invalid combination of knot value count, control point count, and degree.");
```
Comment (sampletext32) [inline src/IxMilia.Converters/Spline2.cs:122]:
```suggestion
                throw new InvalidOperationException("Invalid combination of knot values count, control points count, and degree.");
```

File: src/IxMilia.Converters/Spline2.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                }
+
+                var exepctedValue = values[0];
+                int missingValueCount;
+                if (values.Count < expectedIdenticalKnots)
+                {
+                    // not enough values
+                    missingValueCount = expectedIdenticalKnots - values.Count;
+                }
+                else if (values.Count < expectedIdenticalKnots || !values.All(v => v == exepctedValue))
```
Comment (sampletext32) [inline src/IxMilia.Converters/Spline2.cs]:
```suggestion
                else if (values.Count < expectedIdenticalKnots || values.Any(v => v != exepctedValue))
```

File: src/IxMilia.Converters/SvgPath.cs [hunk 5]
```diff
@@ -152,4 +152,40 @@ public override string ToString()
...
+                "C", // cubic bezier absolute
+                ControlPoint1X.ToDisplayString(),
+                ControlPoint1Y.ToDisplayString(),
+                ",",
+                ControlPoint2X.ToDisplayString(),
+                ControlPoint2Y.ToDisplayString(),
+                ",",
+                EndLocationX.ToDisplayString(),
+                EndLocationY.ToDisplayString()
+            });
```
Comment (sampletext32) [inline src/IxMilia.Converters/SvgPath.cs]: This is kinda heavy operation, better replace it with standart ```string.Format``` call.

**Comment by sampletext32** [comment]
Would you kindly help me with some line.
I need to convert whole DxfFile into Svg. How can I create proper Options for it?

**Comment by sampletext32** [comment]
I have encountered an knot of 1.03 exception with this file.
[buged-file.zip](https://github.com/ixmilia/converters/files/5455073/buged-file.zip)
image

**Comment by brettfo** [comment]
I just added a commit relaxing the range of knot values; let me know what that does for you.

**Comment by sampletext32** [comment]
Still having the same issue. I think it's some-kind bugged dxf file. Everything else seems fine.

**Comment by sampletext32** [comment]
There's actually a catch.
I am investigating this file. 
[zo9yw5e7.zip](https://github.com/ixmilia/converters/files/5469629/zo9yw5e7.zip)

Direct DxfEntities Rendering gives this
image

After conversion to Svg I get this (rendering is done with [vvvv/Svg](https://github.com/vvvv/SVG))
image

Unfortunately can't share any code to help you.

**Comment by brettfo** [comment]
I was inappropriately skipping the first bezier segment in a complex curve.  Using your drawing and the latest commit I was able to generate the following SVG:

image

**Comment by sampletext32** [comment]
Yeah, checked this branch Converter
I am getting this
image
I suppose there's something in ```vvvv/SVG``` rendering, that produces inappropriate connections.

I am using this piece of code to draw an svg.
image


In any case
Definitely better than previous.
The same Svg in browser looks perfect
image

**Comment by brettfo** [comment]
Thanks for the feedback.  Since it's rendering in a browser I'll merge this.

It might be interesting to submit the generated SVG to vvv/SVG so they can see what's going on with their path handling.

**Comment by sampletext32** [comment]
Thanks for your work. I really enjoyed your quick response for my problem.

**Pull request merged**